src/manifolds/SymmetricPositiveDefiniteLinearAffine.jl

```diff
@@ -6,6 +6,146 @@ matrix logarithms and exponentials, which yields a linear and affine metric.
 """
 struct LinearAffineMetric <: RiemannianMetric end
 
+@doc raw"""
+    SPDPoint
+
+Store the result of `eigen(p)` of an SPD matrix and (optionally) ``p^{\frac{1}[2}}`` and ``p^{-\frac{1}[2}}``
+to avoid their repeated computations.
+
+This result only has the result of `eigen` as a mandatory storage, the other three
+can be stored. If they are not stored they are computed and returned (but then still not stored)
+when required.
+
+# Constructor
+
+    SPDPoint(p::AbstractMatrix; store_p=true, store_sqrt=true, store_sqrt_inv=true)
+
+Create an SPD point using an symmetric positive defincite matrix `p`, where you can optionally store `p`, `p_sqrt` and `p_sqrt_inv`
+"""
+struct SPDPoint{
+    P<:Union{<:AbstractMatrix,Missing},
+    Q<:Union{<:AbstractMatrix,Missing},
+    R<:Union{<:AbstractMatrix,Missing},
+    E,
+}
+    p::P
+    eigen::E
+    sqrt::Q
+    sqrt_inv::R
+end
+function SPDPoint(p::AbstractMatrix; store_p=true, store_sqrt=true, store_sqrt_inv=true)
+    e = eigen(Symmetric(p))
+    U = e.vectors
+    S = max.(e.values, floatmin(eltype(e.values)))
+    if store_sqrt
+        s_sqrt = Diagonal(sqrt.(S))
+        p_sqrt = U * s_sqrt * transpose(U)
+    else
+        p_sqrt = missing
+    end
+    if store_sqrt_inv
+        s_sqrt_inv = Diagonal(1 ./ sqrt.(S))
+        p_sqrt_inv = U * s_sqrt_inv * transpose(U)
+    else
+        p_sqrt_inv = missing
+    end
+    if store_p
+        q = p
+    else
+        q = missing
+    end
+    return SPDPoint{typeof(q),typeof(p_sqrt),typeof(p_sqrt_inv),typeof(e)}(
+        q,
+        p_sqrt,
+        p_sqrt_inv,
+    )
+end
+
+@doc raw"""
+    get_point(p::AbstractMatrix)
+    get_point(p::SPDPoint)
+
+return the point `p`, which is either the identity if `p` is an abstract matrix or stored within the [`SPDPoint`](@ref),
+and is reconstructed from `p.eigen` otherwise.
+"""
+get_point(p)
+
+get_point(p::AbstractMatrix) = p
+get_point(p::SPDPoint) = p.p
+get_point(p::SPDPoint{Missing}) = p.eigen.vectors * p.eigen.values * p.eigen.vectors'
+
+@doc raw"""
+    get_p_sqrt(p::AbstractMatrix)
+    get_p_sqrt(p::SPDPoint)
+
+return ``p^{\frac{1}{2}}`` by either computing it (if it is missing or for the `AbstractMatrix`)
+or returning the stored value from within the [`SPDPoint`](@ref).
+"""
+function get_p_sqrt(p::AbstractMatrix)
+    e = eigen(Symmetric(p))
+    U = e.vectors
+    S = max.(e.values, floatmin(eltype(e.values)))
+    Ssqrt = Diagonal(sqrt.(S))
+    return Symmetric(U * Ssqrt * transpose(U))
+end
+get_p_sqrt(p::SPDPoint) = Symmetric(p.sqrt)
+function get_p_sqrt(p::SPDPoint{P,Missing}) where {P}
```

Comment:
```suggestion
function get_p_sqrt(p::SPDPoint{P,Missing}) where {P<:AbstractMatrix}
```
This should remove the ambiguity I think.